ssd1306: Support connecting SPI and I2C simultaneously

soburi opened the pull request:
Support connecting different display for each SPI and I2C at the same time.

In a case like DTS below.

```
&spi1 {
        ssd1306_spi: ssd1306@0 {
		compatible = "solomon,ssd1306fb";
                ...
        };

};

&i2c0 {
        ssd1306_i2c: ssd1306@3c {
		compatible = "solomon,ssd1306fb";
		...
	};
};
```


In addition, I'm doing the necessary correspondence for connecting multiple devices.

soburi committed:
drivers: display: ssd1306: Support connecting SPI and I2C at the same

Support connecting different display for each SPI and I2C
at the same time.

In a case like DTS below.

```
&spi1 {
        ssd1306_spi: ssd1306@0 {
		compatible = "solomon,ssd1306fb";
                ...
        };

};

&i2c0 {
        ssd1306_i2c: ssd1306@3c {
		compatible = "solomon,ssd1306fb";
		...
	};
};
```

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
drivers: display: ssd1306: determin sh1106 by dts compatibility

Determine sh1106 from the `compatibility` value instead of
the SSD1306_CONTROLLER_TYPE setting.

Change the settings in `boards/shields/ssd1306/sh1106_128x64.overlay`
to follow this change.
Remove the SSD1306_CONTROLLER_TYPE from its Kconfig.defconfig,
and set the `compatibility` to `sinowealth,sh1106`.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
drivers: display: ssd1306: replace SSD1306_REVERSE_MODE by property

When multiple devices are connected, the SSD1306_REVERSE_MODE setting
cannot switch for each device.
Add an equivalent setting to the devicetree properties to replace it.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
drivers: display: ssd1306: add dts properties to ssd1306_config

Store properties defined in dts in ssd1306_config's fields.
And replace code that uses DT_INST_PROP (0, ...) by config properties.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

jfischer-no commented on drivers/display/ssd1306.c:
```diff
@@ -20,25 +20,6 @@ LOG_MODULE_REGISTER(ssd1306, CONFIG_DISPLAY_LOG_LEVEL);
 
 #include "ssd1306_regs.h"
 
-#if DT_INST_PROP(0, segment_remap) == 1
-#define SSD1306_PANEL_SEGMENT_REMAP	true
-#else
-#define SSD1306_PANEL_SEGMENT_REMAP	false
-#endif
-
-#if DT_INST_PROP(0, com_invdir) == 1
-#define SSD1306_PANEL_COM_INVDIR	true
-#else
-#define SSD1306_PANEL_COM_INVDIR	false
-#endif
-
-#if DT_INST_PROP(0, com_sequential) == 1
-#define SSD1306_COM_PINS_HW_CONFIG	SSD1306_SET_PADS_HW_SEQUENTIAL
-#else
-#define SSD1306_COM_PINS_HW_CONFIG	SSD1306_SET_PADS_HW_ALTERNATIVE
-#endif
-
-#define SSD1306_PANEL_NUMOF_PAGES	(DT_INST_PROP(0, height) / 8)
```
This patch should be first change to add multi instance support for this driver.

soburi replied:
I reordered commits.

fabiobaltieri replied:
@soburi this is still the last patch, can you check again?

soburi replied:
@fabiobaltieri 
Thank you for pointing out. It's certainly not good.
I fixed it.

jfischer-no commented on drivers/display/ssd1306.c:
```diff
@@ -122,15 +112,9 @@ static int ssd1306_write_bus_spi(const struct device *dev, uint8_t *buf, size_t
-	struct spi_buf_set tx_bufs = {
-		.buffers = &tx_buf,
-		.count = 1
-	};
```
Please stop unnecessary code reformatting.

soburi replied:
Restored it.
And also fix the follows.
https://github.com/zephyrproject-rtos/zephyr/compare/7b2136aabb677b5eec3e34fbaaacd305c2b29c89..48bb5d6f44bf05860258f87acc8b74676742d353#diff-7da6d958478082c2b7f2549d889797958fbf4226bbee0fadbf40e3907479afadL86-R82

jfischer-no commented on drivers/display/Kconfig.ssd1306:
```diff
@@ -35,9 +35,4 @@ config SSD1306_SH1106_COMPATIBLE
 
 endchoice
 
-config SSD1306_REVERSE_MODE
-	bool "SSD1306 reverse mode"
-	help
-	  SSD1306 reverse video mode.
-
```
This commit should be second change before multi instance support commit.

soburi replied:
I reordered commits.

soburi commented on drivers/display/ssd1306.c:
```diff
@@ -452,7 +480,35 @@ static struct display_driver_api ssd1306_driver_api = {
+			      &ssd1306_config_##inst, POST_KERNEL, CONFIG_DISPLAY_INIT_PRIORITY,   \
+			      &ssd1306_driver_api);
+
+DT_INST_FOREACH_STATUS_OKAY(SSD1306_DEFINE)
```
I reordered commits.

soburi commented on drivers/display/ssd1306.c:
```diff
@@ -48,56 +48,75 @@ LOG_MODULE_REGISTER(ssd1306, CONFIG_DISPLAY_LOG_LEVEL);
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
+	struct spi_dt_spec spi;
+#endif
+};
```
Removed #if in SPI and I2C conditionals.
However, the following parts are left because they cause unreferenced warnings.

https://github.com/zephyrproject-rtos/zephyr/pull/56887/files#diff-7da6d958478082c2b7f2549d889797958fbf4226bbee0fadbf40e3907479afadR68
https://github.com/zephyrproject-rtos/zephyr/pull/56887/files#diff-7da6d958478082c2b7f2549d889797958fbf4226bbee0fadbf40e3907479afadR94

fabiobaltieri commented:
@erwango @jfischer-no could you do another review pass once you have a chance?

fabiobaltieri commented on drivers/display/ssd1306.c:
```diff
@@ -56,41 +65,45 @@ struct ssd1306_data {
 	const struct ssd1306_config *config = dev->config;
 
-	return device_is_ready(config->bus.bus);
+	return device_is_ready(config->bus.i2c.bus);
```
You can use `i2c_is_ready_dt` here.

soburi replied:
Fixed it.

soburi commented on drivers/display/ssd1306.c:
```diff
@@ -56,6 +37,16 @@ struct ssd1306_config {
+	bool segment_remap;
+	bool com_invdir;
+	bool com_sequential;
+	int prechargep;
```
Thank you for poiting out.
I change these to uint8_t or uint16_t. (The register values are usually 8bit.)

soburi commented on dts/bindings/display/solomon,ssd1306fb-common.yaml:
```diff
@@ -48,3 +48,7 @@ properties:
+
+  reverse:
+    type: boolean
+    description: Enable display inversion
```
The property name change to `inversion-on`.

soburi commented on boards/shields/ssd1306/Kconfig.defconfig:
```diff
@@ -5,10 +5,6 @@ if SHIELD_SSD1306_128X64 || SHIELD_SSD1306_128X64_SPI || SHIELD_SSD1306_128X32 |
 
 if DISPLAY
 
-choice SSD1306_CONTROLLER_TYPE
-	default SSD1306_SH1106_COMPATIBLE if SHIELD_SH1106_128X64
-endchoice
```
I squashed it into one commit and updated the message .

fabiobaltieri commented:
@jfischer-no can you do another pass?

jfischer-no commented:
> @jfischer-no can you do another pass?

I will look for my displays and test this PR ASAP

erwango commented on dts/bindings/display/solomon,ssd1306fb-common.yaml:
```diff
@@ -49,6 +49,10 @@ properties:
       If connected directly the MCU pin should be configured
       as active low.
 
+  sh1106-compatible:
+    type: boolean
+    description: Indicate device is compatible with SH1106
```
Why not adding a proper and usual compatible property to do that?
That's the point of compatible properties and if required several compatibles can be provided.

soburi replied:
I fixed it detecting with dts compatibility.

soburi committed:
devicetree: rename and publish DT_COMPAT_ON_BUS_INTERNAL

Rename `DT_COMPAT_ON_BUS_INTERNAL` to
`DT_HAS_COMPAT_ON_BUS_STATUS_OKAY` to make it a public DT API.

It is helpful for code that handles multiple DT_DRV_COMPAT in one file,
such as in the following cases.

```
 #if (DT_HAS_COMPAT_ON_BUS_STATUS_OKAY(some_sensor, i2c) || \
      DT_HAS_COMPAT_ON_BUS_STATUS_OKAY(another_sensor, i2c)
 ...
 #endif

 #define DT_DRV_COMPAT some_sensor
 DT_INST_FOREACH_STATUS_OKAY(DEFINE_SOME_SENSOR)

 #undef DT_DRV_COMPAT
 #define DT_DRV_COMPAT another_sensor
 DT_INST_FOREACH_STATUS_OKAY(DEFINE_ANOTHER_SENSOR)
```

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

fabiobaltieri reviewed (dismissed):
General higher level thought: the current approach for driver supporting multiple part numbers by redefining or defining `DT_DRV_COMPAT` conditionally does not feel very scalable. In Linux, where everything happens in runtime, there are these big compatible to id and then id to feature/quirk table: https://lxr.missinglinkelectronics.com/linux/drivers/iio/imu/st_lsm6dsx/st_lsm6dsx_i2c.c#L39 https://lxr.missinglinkelectronics.com/linux/drivers/iio/imu/st_lsm6dsx/st_lsm6dsx_core.c#L91. We may probably need something to close the gap down the road if we want to support more variants... Maybe @galak has some ideas already. :-)

jfischer-no commented:
> General higher level thought: the current approach for driver supporting multiple part numbers by redefining or defining `DT_DRV_COMPAT` conditionally does not feel very scalable. In Linux, where everything happens in runtime, there are these big compatible to id and then id to feature/quirk table: https://lxr.missinglinkelectronics.com/linux/drivers/iio/imu/st_lsm6dsx/st_lsm6dsx_i2c.c#L39 https://lxr.missinglinkelectronics.com/linux/drivers/iio/imu/st_lsm6dsx/st_lsm6dsx_core.c#L91. We may probably need something to close the gap down the road if we want to support more variants... Maybe @galak has some ideas already. :-)

What is in the way to do it? We also have examples in the tree, e.g. usb_dc_dw.c

jfischer-no commented on boards/shields/ssd1306/sh1106_128x64.overlay:
```diff
@@ -25,5 +25,6 @@
 		segment-remap;
 		com-invdir;
 		prechargep = <0x22>;
+		sh1106-compatible;
```
Please remove `sh1106-compatible;` property.

```
-- Found devicetree overlay: /home/jofi/zephyr-project/zephyr/boards/shields/ssd1306/sh1106_128x64.overlay
devicetree error: 'sh1106-compatible' appears in /soc/i2c@40066000/ssd1306@3c in /home/jofi/zephyr-project/zephyr/samples/subsys/display/cfb/outdir/zephyr/zephyr.dts.pre, but is not declared in 'properties:' in /home/jofi/zephyr-project/zephyr/dts/bindings/display/sinowealth,sh1106-i2c.yaml
```

soburi replied:
Thank you for pointing out.
It's my careless mistake.
I confirmed that the compilation passed with the following command.

```
 west build -b frdm_k64f -d ../k64f samples/subsys/display/lvgl/ -- -DSHIELD=sh1106_128x64
```

soburi commented:
@fabiobaltieri 
Even it is not an essential solution, I clean up it so that undef is not used.

There are some drawbacks, but I think the macros that resolve devicetree work well currently and it is solid infrastructure.
I would like to follow the overall policy on this point.

BTW, https://github.com/zephyrproject-rtos/zephyr/issues/22941 Is this active? Should I better  to raise the issue?

@jfischer-no 

I'm sorry it's so often.
Could you please take a look at it again?

fabiobaltieri commented:
> @fabiobaltieri Even it is not an essential solution, I clean up it so that undef is not used.
> 
> There are some drawbacks, but I think the macros that resolve devicetree work well currently and it is solid infrastructure. I would like to follow the overall policy on this point.

Oh yeah, sorry I wasn't implying you should work on that here, just thinking that cases like these may benefit from something at device driver level. I like the change though, thanks.

fabiobaltieri commented:
> @fabiobaltieri What is in the way to do it? We also have examples in the tree, e.g. usb_dc_dw.c

Hey only seen this now: yeah I guess it's one way of doing it but I'd say it does not scale as well compared to, for example, this: https://elixir.bootlin.com/linux/latest/source/drivers/gpio/gpio-pca953x.c#L1319

mbolivar-ampere commented on include/zephyr/devicetree.h:
```diff
@@ -3755,6 +3755,40 @@
+ * @return 1 if any enabled node with that compatible is on that bus type,
+ *         0 otherwise
+ */
+#define DT_COMPAT_ANY_ON_BUS(compat, bus) \
```
Please rename this to DT_HAS_COMPAT_ON_BUS_STATUS_OKAY for consistency with other APIs like DT_HAS_COMPAT_STATUS_OKAY.

soburi replied:
I renamed it to DT_*ANY*_HAS_COMPAT_ON_BUS_STATUS_OKAY, considering symmetry with DT_ANY_INST_ON_BUS_STATUS_OKAY.
The only difference is that the argument is instance number or compatibility, so I changed it accordingly.

mbolivar-ampere replied:
I think the symmetry with `DT_ANY_INST` is not clear. "ANY_INST" means "any instance of DT_DRV_COMPAT". "ANY_HAS" does not make sense to me.

soburi replied:
I reconsidered, and I got it you make sense.
I applied it, Thank you.

mbolivar-ampere commented on include/zephyr/devicetree.h:
```diff
@@ -3755,6 +3755,40 @@
 	DT_STRING_UNQUOTED_OR(DT_DRV_INST(inst), name, default_value)
 
+/**
+ * @brief Test if any given compat is on a bus of a given type
+ *        and has status okay
+ *
+ * In usually, the DT_ANY_INST_ON_BUS_STATUS_OKAY can use for check for node
+ * that is on specified bus existence check.
+ * This is useful for handle multiple compat in single source file.
+ *
+ * Example devicetree overlay:
+ *
+ * @code{.dts}
+ *     &i2c0 {
+ *            temp: temperature-sensor@76 {
+ *                     compatible = "vnd,some-sensor";
+ *                     reg = <0x76>;
+ *            };
+ *     };
+ * @endcode
+ *
+ * Example usage, assuming `i2c0` is an I2C bus controller node, and
+ * therefore `temp` is on an I2C bus:
+ *
+ * @code{.c}
+ *     DT_COMPAT_ANY_ON_BUS(vnd_some_sensor, i2c) // 1
+ * @endcode
+ *
+ * @param compat lowercase-and-underscores compatible, without quotes
+ * @param bus a binding's bus type as a C token, lowercased and without quotes
+ * @return 1 if any enabled node with that compatible is on that bus type,
+ *         0 otherwise
```
Here are some suggested documentation changes:

```c
/*
 * @brief Test if any enabled node with the given compatible is on
 *        the given bus type
 *
 * This is like DT_ANY_INST_ON_BUS_STATUS_OKAY(), except it can also
 * be useful for handling multiple compatibles in single source file.
 *
 * Example devicetree overlay:
 *
 * @code{.dts}
 *     &i2c0 {
 *            temp: temperature-sensor@76 {
 *                     compatible = "vnd,some-sensor";
 *                     reg = <0x76>;
 *            };
 *     };
 * @endcode
 *
 * Example usage, assuming `i2c0` is an I2C bus controller node, and
 * therefore `temp` is on an I2C bus:
 *
 * @code{.c}
 *     DT_HAS_COMPAT_ON_BUS_STATUS_OKAY(vnd_some_sensor, i2c) // 1
 * @endcode
 *
 * @param compat lowercase-and-underscores compatible, without quotes
 * @param bus a binding's bus type as a C token, lowercased and without quotes
 * @return 1 if any enabled node with that compatible is on that bus type,
 *         0 otherwise
 */
```

soburi replied:
Thank you for improvement.
Applied it.

mbolivar-ampere reviewed (approved):
Thanks for your patch. I agree we can expose this macro in devicetree.h. I have a couple of minor remarks.

mbolivar-ampere reviewed (changes requested):
Sorry, I meant to request the changes in the previous review, not approve. I mistakenly clicked 'approve'.

soburi commented:
@erwango 
Could you take look at please?

fabiobaltieri commented:
@erwango @mbolivar-ampere ping

erwango reviewed (dismissed):
Ok for shield part.

jfischer-no commented:
What changed again?

soburi commented:
@jfischer-no 

> What changed again?

Sinse Aug 9,
I fixed about pointed by @mbolivar-ampere https://github.com/zephyrproject-rtos/zephyr/pull/56887#pullrequestreview-1578986144

and I change to remove `#undef DT_DRV_COMPAT` line.

https://github.com/zephyrproject-rtos/zephyr/compare/382db6dd259f314d9527b2e25eedd2aa303a0c37..e4646cf81d7346a04b62e7497c8db786b5533297

soburi commented:
@mbolivar-ampere
Could you take a look again, please?

soburi commented:
@mbolivar-ampere
ping
